### PR TITLE
Add AMMClawback Transaction (XLS-0073d)

### PR DESCRIPTION
**Pull request by yinyiqian1**
### Context of Change
Spec link: https://github.com/XRPLF/XRPL-Standards/discussions/212

This PR includes:
1. Allow AMMCreate for clawback-enabled tokens
    changes are in `AMMCreate` and `AMM_test`
2. AMMDeposit prohibits depositing if either of the paired tokens is frozen
    changes are in `AMMDeposit` and `AMM_test` 
3. Added AMMClawback transaction.
    mainly in `AMMClawback` and `AMMClawback_test`

___________________________________________________________________________
Some refactor happens because I want to call some functions from AMMWithdraw.
**1. Refactor of `withdraw`:**
- Added static `withdraw` function to return the `amountWithdrawActual` and `amount2WithdrawActual` as well.
- private overload `withdraw`.
- Previously in `withdraw` function, it checks `if (!(ctx_.tx[sfFlags] & (tfWithdrawAll | tfOneAssetWithdrawAll)))` to decide if it needs to call `adjustAmountsByLPTokens`. Because these flags only belong to `AMMWithdraw`, to make this function also work for AMMClawback, I changed it by passing enum WithdrawAll.

**2. Refactor of `equalWithdrawTokens`:**
  Because I want to call this function in `AMMClawback`, I make a public static `equalWithdrawTokens` function. And also private overload `equalWithdrawTokens`

**3. Added `deleteAMMAccountIfEmpty`**
  Since I need to reuse the logic in AMMWithdraw that when all the tokens are withdrawn, the amm account is deleted, so I added function `deleteAMMAccountIfEmpty`.

___________________________________________________________________________
**Reason of adding flag `tfClawTwoAssets`:**
If the issuer issues both assets `A` and `B` in the amm pool, when clawing back `A`, `tfClawTwoAssets` is used to decide if the paired token `B` to be clawed back as well or goes back to the holder. If  tfClawTwoAssets` is set true, the paired token will be clawed back as well. Otherwise, the paired token goes back to the holder.

**Reason of adding both `asset` and `amount` in the `AMMClawback` request:**
`amount` is used to decide if the issuer wants to clawback all the tokens or not. (if `amount` is not provided, then clawback all)  Imagine if an issuer issues both asset `A` and `B` in the amm pool, and the issuer wants to clawback all `A` and give back all the corresponding `B` to the holder. Then the issuer just needs to specify `A` in `asset` field and leave out `amount` field.

**Reasons for setting tfee as 0:**
Because we are doing a two-asset withdrawal. tfee is not used.

**Reasons for using asset1 and asset2 to navigate amm**
To keep the same with other amm transactions and make it simple.
So we require providing asset1 and asset2 in the AMMClawback request.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [x] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [x] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

**Comment by mvadari** [comment]
Note: The PR title is wrong, it should be XLS-73d

**Comment by yinyiqian1** [comment]
> Note: The PR title is wrong, it should be XLS-73d
Just corrected it. Thank you.

File: include/xrpl/protocol/TER.h [hunk 1]
```diff
@@ -139,6 +139,8 @@ enum TEMcodes : TERUnderlyingType {
 
     temARRAY_EMPTY,
     temARRAY_TOO_LARGE,
+    temBAD_ASSET_AMOUNT,
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply include/xrpl/protocol/TER.h:142]: There's a temBAD_AMOUNT already defined for non-positive amount errors.

File: src/xrpld/app/tx/detail/AMMDeposit.cpp [hunk 2]
```diff
@@ -244,6 +244,36 @@ AMMDeposit::preclaim(PreclaimContext const& ctx)
...
 
+    if (ctx.view.rules().enabled(featureAMMClawback))
+    {
+        // Check if either of the assets is frozen, AMMDeposit is not allowed
+        // if either asset is frozen
+        auto checkAsset = [&](std::optional<Issue> const& asset) -> TER {
+            if (!asset.has_value())
+                return temMALFORMED;  // LCOV_EXCL_LINE
+
+            if (isFrozen(ctx.view, accountID, asset.value()))
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMDeposit.cpp:255]: `isFrozen` checks both global and individual freeze

File: include/xrpl/protocol/TER.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gregtatcam) [reply include/xrpl/protocol/TER.h:142]: That's what I mean, don't need to define new ones.

[the remark this answers is not in the document]
Comment (yinyiqian1) [reply include/xrpl/protocol/TER.h:142]: Then the request's error message for this one will be inaccurate: `Can only send positive amounts.`
But I intended to return error message as `Malformed: Amount does not match Asset.`

[the remark this answers is not in the document]
Comment (gregtatcam) [reply include/xrpl/protocol/TER.h:142]: could just use `temMALFORMED` then.

[the remark this answers is not in the document]
Comment (gregtatcam) [reply include/xrpl/protocol/TER.h:142]: But I'm fine if you feel strongly about using new codes.

[the remark this answers is not in the document]
Comment (mvadari) [reply include/xrpl/protocol/TER.h:142]: Why not change the error message? IIRC that's not considered a breaking change.

File: src/xrpld/app/tx/detail/AMMWithdraw.h [hunk 3]
```diff
@@ -80,33 +80,6 @@ class AMMWithdraw : public Transactor
...
-     * @param amountBalance
-     * @param amountWithdraw
-     * @param amount2Withdraw
-     * @param lpTokensAMMBalance current AMM LPT balance
-     * @param lpTokensWithdraw
-     * @param tfee
-     * @return
-     */
-    std::pair<TER, STAmount>
-    withdraw(
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: can we keep withdrawAll as bool? Otherwise, it is kind of verbose to use
`WithdrawAll withdrawAll = ctx_.tx[sfFlags] & (tfWithdrawAll | tfOneAssetWithdrawAll) ? WithdrawAll::Yes : WithdrawAll::No;`
instead of `bool withdrawAll = ctx_.tx[sfFlags] & (tfWithdrawAll | tfOneAssetWithdrawAll)`

File: include/xrpl/protocol/TER.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply include/xrpl/protocol/TER.h:142]: > Why not change the error message? IIRC that's not considered a breaking change.

the temBAD_AMOUNT is used in other places to check positive amounts. So here I need a new error type. That's why I created a new error type temBAD_ASSET_AMOUNT. The error message is bind to each error type.

File: src/xrpld/app/tx/detail/AMMWithdraw.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: It's just more clear when you call this function with enum as opposed to true/false. Perhaps you can add a helper function, which takes the flags and returns enum to make it less verbose?

[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: Also, in `withdraw` function, it checks the `lpTokens` hold by the `account`.
`auto const lpTokens = ammLPHolds(view, ammSle, account, journal);` 
If we pass `issuer` as `account`, then it gets the issuer's lptokens, which will affect the following logic which uses this `lpTokens` variable. For example, the issuer does not hold any lptokens, and we possibly will return as `lpTokensWithdrawActual > lpTokens`.

[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: Good point. What if the `withdrawAll` is extended: `enum WithdrawAll {Yes, TransferToIssuer, TransferAllToIssuer, No}`? Anything but `No` means withdrawAll. If `TransferToIssuer` is set then `amount` is transferred to the issuer and if `TransferAllToIssuer` is set then `amount2` is transferred to the issuer. The issuer account doesn't have to be included since it's included in the amounts.

File: include/xrpl/protocol/TER.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvadari) [reply include/xrpl/protocol/TER.h:142]: It's also thrown in the escrow code if the asset isn't XRP.

File: src/xrpld/app/tx/detail/AMMWithdraw.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: Yeah, we can make this change

[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: @gregtatcam
maybe we can rename `WithdrawAll` to `WithdrawOption`?
`enum WithdrawOption {Yes, TransferToIssuer, TransferAllToIssuer, No}`

I have another concern,
Since withdraw is used in many places, is it better to keep its function as simple as possible? And clawback change is on top of this layer by using its return value.
Of course, by adding logics directly into the withdraw function is another option and makes less code change in caller functions. But this will make `withdraw` more complicated. And make it check the `WithdrawOption` each time. What do you think?

[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: Renaming sounds good to me.

We either have a common function or have two separate functions. The latter is not an option. Since the common function has to work for both use-cases, it does become more complicated one way or another. To mitigate this complexity in AMMWithdraw, we are adding a private non-static `withdraw`, which is going to hide most of the changes from the callers. I'm ok with returning amounts, again given the overload, if you feel strongly about it. Though I would argue that transferring to the holder and then to the issuer is not ideal either.

File: src/test/app/AMM_test.cpp [hunk 4]
```diff
@@ -6862,13 +6897,140 @@ struct AMM_test : public jtx::AMMTest
...
+        testcase("test AMMDeposit with frozen assets");
+        using namespace jtx;
+
+        // This lambda function is used to create trustlines
+        // between gw and alice, and create and return an AMM account.
+        auto setupAMM = [&](Env& env) -> AMM* {
+            env.fund(XRP(1'000), gw);
+            fund(env, gw, {alice}, XRP(1'000), {USD(1'000)}, Fund::Acct);
+            env.close();
+            AMM* amm = new AMM(env, alice, XRP(100), USD(100), ter(tesSUCCESS));
```
Comment (gregtatcam) [inline src/test/app/AMM_test.cpp:6959]: Please use `shared_ptr` or add a callback to the lambda:

    auto setupAMM = [&](Env& Env, std::function<void(AMM& amm)> cb)

File: src/test/app/AMMClawback_test.cpp [hunk 5]
```diff
@@ -0,0 +1,1444 @@
...
+            // gw issues 100 USD to Alice.
+            auto const USD = gw["USD"];
+            env.trust(USD(1000), alice);
+            env(pay(gw, alice, USD(100)));
+            env.close();
+
+            AMM amm(env, gw, XRP(100), USD(100), ter(tesSUCCESS));
+
+            // Issuer can not clawback from himself.
+            env(amm::ammClawback(gw, gw, USD, XRP, std::nullopt, std::nullopt),
```
Comment (gregtatcam) [inline src/test/app/AMMClawback_test.cpp:94]: Similarly should test holder/holder.

File: src/test/app/AMMClawback_test.cpp [hunk 6]
```diff
@@ -0,0 +1,1444 @@
...
+                    gw2, alice, EUR, USD, std::nullopt, std::nullopt),
+                ter(tesSUCCESS));
+            env.close();
+
+            env.require(balance(carol, gw2["EUR"](2750)));
+            env.require(balance(carol, gw["USD"](4000)));
+            BEAST_EXPECT(!amm.ammExists());
+        }
+
+        // Test AMMClawback for USD/XRP pool. The assets are issued by different
```
Comment (gregtatcam) [inline src/test/app/AMMClawback_test.cpp:1003]: Typo? "The assets are issued by different issuer". It's XRP/USD.

File: src/test/jtx/impl/AMM.cpp [hunk 7]
```diff
@@ -823,6 +823,29 @@ pay(Account const& account, AccountID const& to, STAmount const& amount)
     jv[jss::Flags] = tfUniversal;
     return jv;
 }
+
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/test/jtx/impl/AMM.cpp:826]: asset and asset2 can be pre-populated. But the other values need to be passed in.
But there's some test case which need to pass in the wrong asset purposely.
The ammclawback does not have multiple situations like ammwithdraw which needs different combinations of parameters. So can we keep it simple as a helper function?

File: src/xrpld/app/tx/detail/AMMWithdraw.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: If we continue use `WithdrawOption`, we have to separate it from `WithdrawAll`. 
`enum WithdrawOption {BothToAccount, AmountToIssuer, BothToIssuer}` is to control if sending to issuer. `enum WithdrawAll {true, false}` is to control if withdraw all.
Because the following situations will be a 6 combinations of the usage:
1. withdraw all to account
2. not withdraw all to account
3. withdraw all + token A to issuer + token B to account
4. not withdraw all + token A to issuer + token B to account
5. withdraw all  + token A and B to issuer
6. not withdraw all + token A and B to issuer

This will make the function taking another parameter and more complex. And all the callers need to be very clear about which flow it's going to take.
Basically for withdraw function, I think it would be better if it just take care of the withdraw logics like: "withdraw to an account."
And returning the withdrawn amounts will be more extensible in the future if these amounts are needed again in another feature.

File: src/libxrpl/protocol/TER.cpp [hunk 8]
```diff
@@ -205,6 +205,8 @@ transResults()
         MAKE_ERROR(temXCHAIN_BRIDGE_BAD_REWARD_AMOUNT, "Malformed: Bad reward amount."),
         MAKE_ERROR(temARRAY_EMPTY,               "Malformed: Array is empty."),
         MAKE_ERROR(temARRAY_TOO_LARGE,           "Malformed: Array is too large."),
+        MAKE_ERROR(temBAD_ASSET_AMOUNT,          "Malformed: Amount does not match Asset."),
+        MAKE_ERROR(temBAD_ASSET_ISSUER,          "Malformed: Issuer does not match Asset."),
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/libxrpl/protocol/TER.cpp:209]: can I keep this temBAD_ASSET_ISSUER? because it stands for issuer does not match asset.

File: src/xrpld/app/tx/detail/AMMWithdraw.cpp [hunk 9]
```diff
@@ -406,22 +428,12 @@ AMMWithdraw::applyGuts(Sandbox& sb)
     if (result != tesSUCCESS)
         return {result, false};
 
-    bool updateBalance = true;
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.cpp:409]: is it more like a helper function? Since `deleteAMMAccount` is in `AMMUtils`, `deleteAMMAccountIfEmpty` is doing one more step by checking if the lpTokenBalance is 0.

[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.cpp:409]: since `deleteAMMAccount` is in AMMUtils, I think `deleteAMMAccountIfEmpty` is just doing one more step by checking if the lptokenBalance is 0. So it looks like a helper function. Can we keep it in utils?

File: src/xrpld/app/tx/detail/AMMWithdraw.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:100]: Sounds good. But let's have a simplified private overload for other AMMWithdraw functions to call.

File: src/xrpld/app/tx/detail/AMMWithdraw.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMWithdraw.cpp:409]: It's updating the view, which with a few exceptions is done in the transactors or View.

File: src/test/app/AMMClawback_test.cpp [hunk 10]
```diff
@@ -0,0 +1,1449 @@
...
+        FeatureBitset const all{jtx::supported_amendments()};
+        testInvalidRequest(all);
+        testFeatureDisabled(all - featureAMMClawback);
+        testAMMClawbackSpecificAmount(all);
+        testAMMClawbackExceedBalance(all);
+        testAMMClawbackAll(all);
+        testAMMClawbackSameIssuerAssets(all);
+        testAMMClawbackSameCurrency(all);
+        testAMMClawbackIssuesEachOther(all);
+        testNotHoldingLptoken(all);
```
Comment (dangell7) [inline src/test/app/AMMClawback_test.cpp:1444]: There are no tests when the asset/s are frozen. We should have tests where the asset was first globally/individually frozen and then clawed back.

Comment (dangell7) [reply src/test/app/AMMClawback_test.cpp:1444]: We should also include a test where 3 holders deposit, then global freeze, then clawed back one by one. (Might be overkill)

File: src/xrpld/app/tx/detail/AMMClawback.cpp [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    auto const asset = ctx.tx[sfAsset];
+
+    if (isXRP(asset))
+        return temMALFORMED;
+
+    if (asset.account != issuer)
+    {
+        JLOG(ctx.j.trace()) << "AMMClawback: Asset's account does not "
+                               "match Account field.";
+        return temBAD_ASSET_ISSUER;
```
Comment (dangell7) [inline src/xrpld/app/tx/detail/AMMClawback.cpp:68]: nit (Optional): It would be great if we could make these agnostic. I use them in another PR, and other PR's to this point have used temMALFORMED/BAD_AMOUNT. I think its a valuable addition. In general the error is (the tx issuer does not match the ledger entry issuer).

File: src/xrpld/app/tx/detail/AMMClawback.cpp [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+    auto const flags = ctx_.tx.getFlags();
+    if (flags & tfClawTwoAssets)
+
+        return rippleCredit(sb, holder, issuer, *amount2Withdraw, true, j_);
+
+    return tesSUCCESS;
+}
+
+std::tuple<TER, STAmount, STAmount, std::optional<STAmount>>
+AMMClawback::equalWithdrawMatchingOneAmount(
```
Comment (dangell7) [inline src/xrpld/app/tx/detail/AMMClawback.cpp:232]: refactor (optional): I believe this function could be refactored/removed. The statement (lpTokensWithdraw > holdLPtokens) is the same as (!clawAmount). So you could handle this all in applyGuts without another function.

File: src/xrpld/app/tx/detail/AMMClawback.cpp [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    auto const accountSle = sb.read(keylet::account(ammAccount));
+    if (!accountSle)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const expected = ammHolds(
+        sb,
+        *ammSle,
+        asset,
+        asset2,
+        FreezeHandling::fhZERO_IF_FROZEN,
```
Comment (dangell7) [inline src/xrpld/app/tx/detail/AMMClawback.cpp:155]: I think there will be an issue here if the trustline is frozen. Wont the AMM return 0? Add tests for global and line freeze to confirm.

**Review by dangell7 — CHANGES_REQUESTED** [review]

File: src/test/app/AMM_test.cpp (hunk 4 above)
Comment (gregtatcam) [reply src/test/app/AMM_test.cpp:6960]: This still needs to be addressed.

File: src/xrpld/app/tx/detail/AMMWithdraw.h [hunk 14]
```diff
@@ -80,54 +83,98 @@ class AMMWithdraw : public Transactor
...
+     * @param amount2Balance current LP asset2 balance
+     * @param lptAMMBalance current AMM LPT balance
+     * @param lpTokens current LPT balance
+     * @param lpTokensWithdraw amount of tokens to withdraw
+     * @param tfee trading fee in basis points
+     * @param withdrawAll if withdrawing all lptokens
+     * @return
+     */
+    static std::tuple<TER, STAmount, STAmount, std::optional<STAmount>>
+    equalWithdrawTokens(
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h:102]: Yes. Added in the recent commit. And also passing in FreezeHandling type because AMMClawback needs to set it to fhIGNORE_FREEZE

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 13 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp]: Yes. Changed to fhIGNORE_FREEZE in recent commit. And also made related changes and added unit tests for global and individual freeze.

File: src/xrpld/app/misc/detail/AMMUtils.cpp [hunk 15]
```diff
@@ -16,12 +16,14 @@
     OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
 //==============================================================================
+#include <xrpld/app/misc/AMMHelpers.h>
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/misc/detail/AMMUtils.cpp]: removed

File: src/xrpld/app/misc/detail/AMMUtils.cpp [hunk 16]
```diff
@@ -16,12 +16,14 @@
...
 #include <xrpld/app/misc/AMMUtils.h>
 #include <xrpld/ledger/Sandbox.h>
 #include <xrpl/basics/Log.h>
 #include <xrpl/protocol/AMMCore.h>
 #include <xrpl/protocol/STAccount.h>
 #include <xrpl/protocol/STObject.h>
+#include <xrpl/protocol/TxFlags.h>
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/misc/detail/AMMUtils.cpp]: removed

File: src/xrpld/app/tx/detail/AMMWithdraw.h [hunk 17]
```diff
@@ -80,54 +83,98 @@ class AMMWithdraw : public Transactor
...
+     * @param lpTokensAMMBalance current AMM LPT balance
+     * @param lpTokensWithdraw amount of lptokens to withdraw
      * @return
      */
     std::pair<TER, STAmount>
-    equalWithdrawTokens(
+    withdraw(
         Sandbox& view,
         AccountID const& ammAccount,
+        AccountID const& account,
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h]: removed

File: src/test/app/AMM_test.cpp (hunk 4 above)
Comment (yinyiqian1) [reply src/test/app/AMM_test.cpp]: updated

File: src/test/app/AMMClawback_test.cpp (hunk 10 above)
Comment (yinyiqian1) [reply src/test/app/AMMClawback_test.cpp:1730]: added in the recent commit

File: src/test/app/AMMClawback_test.cpp (hunk 6 above)
Comment (yinyiqian1) [reply src/test/app/AMMClawback_test.cpp]: updated

File: src/test/app/AMMClawback_test.cpp (hunk 5 above)
Comment (yinyiqian1) [reply src/test/app/AMMClawback_test.cpp:94]: updated

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 12 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:232]: then the code will have to declare lpTokensWithdraw and amount2Withdraw in a common place.
And we will have another `if (clawAmount)` to calculate these values. Although in the following check `if (!clawAmount || lpTokensWithdraw > holdLPTokens)`'s body, these values are not used.
```
    STAmount lpTokensWithdraw;
    Number amount2Withdraw;
    if (clawAmount)
    {
        auto frac = Number{*clawAmount} / amountBalance;
        amount2Withdraw = amount2Balance * frac;
        lpTokensWithdraw = toSTAmount(lptAMMBalance.issue(), lptAMMBalance * frac);
    }
    if (!clawAmount || lpTokensWithdraw > holdLPTokens)
           // withdraw all
    else
           // withdraw

```
I prefer to keep the code logic together instead of declaring some variable which can only have value if `if (clawAmount)` is true. And those variables are used in the following `if condition` and `else body`.

File: src/xrpld/app/tx/detail/AMMClawback.cpp [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+    auto const ter = rippleCredit(sb, holder, issuer, amountWithdraw, true, j_);
+    if (ter != tesSUCCESS)
+        return ter;  // LCOV_EXCL_LINE
+
+    // if the issuer issues both assets and sets flag tfClawTwoAssets, we
+    // will claw the paired asset as well. We already checked if
+    // tfClawTwoAssets is enabled, the two assets have to be issued by the
+    // same issuer.
+    auto const flags = ctx_.tx.getFlags();
+    if (flags & tfClawTwoAssets)
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:225]: I'll add `if (!amount2Withdraw) return tecINTERNAL;`. because it should have value, otherwise, it's an internal error. What do you think?

File: src/xrpld/app/tx/detail/AMMWithdraw.h [hunk 19]
```diff
@@ -80,6 +84,75 @@ class AMMWithdraw : public Transactor
...
+     * @param tfee trading fee in basis points
+     * @param withdrawAll if withdraw all lptokens
+     * @return
+     */
+    static std::tuple<TER, STAmount, STAmount, std::optional<STAmount>>
+    withdraw(
+        Sandbox& view,
+        AccountID const& ammAccount,
+        AccountID const& account,
+        SLE const& ammSle,
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMWithdraw.h]: updated

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 12 above)
Comment (gregtatcam) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:235]: This is just a simple check with no need for the temp variables:
 
    if (!clawAmount || (lptAMMBalance * *clawAmount / amountBalance) > holdLPTokens)

If you decide to keep the function then makes sense to pass in `holdLPTokens` since it's already fetched.

Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:235]: make sense, I'll pass in holdLPTokens

**Review by gregtatcam — APPROVED** [review]
👍 LGTM

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 11 above)
Comment (dangell7) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:68]: I was mistaken on this. This is not checking the ledger index. Because its not, and this is technically just a MALFORMED txn I would suggest we use malformed here. We dont have an infinite number of error codes and so creating 2 new error codes just for this where we have used MALFORMED in the past isnt a good practice imo.

**Comment by dangell7** [comment]
Just to be clear in an AMM where USD/EUR it is not possible to only claw back the EUR correct?

**Review by dangell7 — CHANGES_REQUESTED** [review]

File: src/xrpld/app/tx/detail/AMMClawback.cpp [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+    auto const [amountBalance, amount2Balance, lptAMMBalance] = *expected;
+
+    TER result;
+    STAmount newLPTokenBalance;
+    STAmount amountWithdraw;
+    std::optional<STAmount> amount2Withdraw;
+
+    auto const holdLPtokens = ammLPHolds(sb, *ammSle, holder, j_);
+    if (holdLPtokens == beast::zero)
+        return tecINTERNAL;
```
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:169]: `tecAMM_EMPTY` is used when the whole AMM's lptoken balance is zero. But `holdLPTokens` is the holder's lptoken share in the amm pool.
If we need some accurate error for the client, we need to define a new one to indicate that the holder does not hold any shares in the AMM pool.

[the remark this answers is not in the document]
Comment (dangell7) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:169]: Sorry `tecAMM_BALANCE` is what I meant.

https://github.com/XRPLF/rippled/blob/bf4a7b6ce8bffd0e098831d95d056e70a433ccce/src/xrpld/app/tx/detail/AMMWithdraw.cpp#L262

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 11 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp:68]: I can change both of the `temBAD_ASSET_AMOUNT`  and `temBAD_ASSET_ISSUER` to `temMALFORMED`. In the beginning, I wanted to provide the accurate error message to the client so that the client knows what was wrong with the request.

File: include/xrpl/protocol/TER.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mvadari) [reply include/xrpl/protocol/TER.h:142]: Why not just change the error message, so that it makes sense both in Escrow and here, instead of needing a new error code? I believe changing the error message isn't considered a breaking change/needing an amendment.

[the remark this answers is not in the document]
Comment (yinyiqian1) [reply include/xrpl/protocol/TER.h:142]: I can change the message for `temBAD_AMOUNT` to "bad amount provided". What do you think?

[the remark this answers is not in the document]
Comment (dangell7) [reply include/xrpl/protocol/TER.h:142]: `Malformed: Bad amount.` would keep it in line with the others.

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp]: updated in the recent commit

File: src/xrpld/app/tx/detail/AMMClawback.cpp (hunk 11 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMClawback.cpp]: updated in the recent commit to `temMALFORMED`

File: include/xrpl/protocol/TER.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yinyiqian1) [reply include/xrpl/protocol/TER.h]: updated in the recent commit

**Review by dangell7 — APPROVED** [review]

File: src/xrpld/app/tx/detail/AMMDeposit.cpp [hunk 21]
```diff
@@ -248,21 +248,33 @@ AMMDeposit::preclaim(PreclaimContext const& ctx)
...
+                return tecFROZEN;
             }
 
-            return true;
+            return tesSUCCESS;
         };
 
-        if (!checkAsset(ctx.tx[sfAsset]) || !checkAsset(ctx.tx[sfAsset2]))
-            return tecFROZEN;
+        // if (!checkAsset(ctx.tx[sfAsset]) || !checkAsset(ctx.tx[sfAsset2]))
```
Comment (gregtatcam) [inline src/xrpld/app/tx/detail/AMMDeposit.cpp:271]: Don't need these comments.

File: src/xrpld/app/tx/detail/AMMDeposit.cpp [hunk 22]
```diff
@@ -248,21 +248,33 @@ AMMDeposit::preclaim(PreclaimContext const& ctx)
...
+        //     return tecFROZEN;
+        if (auto const ter = checkAsset(ctx.tx[sfAsset]))
```
Comment (gregtatcam) [inline src/xrpld/app/tx/detail/AMMDeposit.cpp:273]: Should add a unit-test to check for this updated condition.

File: src/test/app/AMM_test.cpp [hunk 23]
```diff
@@ -1013,8 +1051,8 @@ struct AMM_test : public jtx::AMMTest
             ammAlice.deposit(bob, XRP(10));
             ammAlice.deposit(
                 bob,
-                XRP(1'000),
-                std::nullopt,
+                gw["USD"](100),
```
Comment (gregtatcam) [inline src/test/app/AMM_test.cpp:1054]: Shouldn't we keep the original test with XRP and null USD?

File: src/test/app/AMM_test.cpp [hunk 24]
```diff
@@ -7030,55 +7068,57 @@ struct AMM_test : public jtx::AMMTest
     run() override
     {
         FeatureBitset const all{jtx::supported_amendments()};
-        testInvalidInstance();
-        testInstanceCreate();
+        // testInvalidInstance();
```
Comment (gregtatcam) [inline src/test/app/AMM_test.cpp:7071]: You commented out most of the tests.

File: src/test/app/AMM_test.cpp (hunk 23 above)
Comment (yinyiqian1) [reply src/test/app/AMM_test.cpp]: accidentally changed it. just updated it back

File: src/xrpld/app/tx/detail/AMMDeposit.cpp (hunk 22 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMDeposit.cpp:271]: just updated. (yesterday it's not ready yet)

File: src/test/app/AMM_test.cpp (hunk 24 above)
Comment (yinyiqian1) [reply src/test/app/AMM_test.cpp]: updated

File: src/xrpld/app/tx/detail/AMMDeposit.cpp (hunk 21 above)
Comment (yinyiqian1) [reply src/xrpld/app/tx/detail/AMMDeposit.cpp]: updated

File: include/xrpl/protocol/Feature.h [hunk 25]
```diff
@@ -80,7 +80,7 @@ namespace detail {
 // Feature.cpp. Because it's only used to reserve storage, and determine how
 // large to make the FeatureBitset, it MAY be larger. It MUST NOT be less than
 // the actual number of amendments. A LogicError on startup will verify this.
-static constexpr std::size_t numFeatures = 79;
+static constexpr std::size_t numFeatures = 80;
```
Comment (shawnxie999) [inline include/xrpl/protocol/Feature.h:83]: should be incremented to 81 after new amendment is merged into develop branch

**Comment by ximinez** [comment]
@yinyiqian1 Are we happy with the commit message being just
```
Add AMMClawback Transaction (XLS-0073d) (#5142)
```
?

If not, please suggest the details you'd like to include.

**Comment by yinyiqian1** [comment]
> @yinyiqian1 Are we happy with the commit message being just
> 
> ```
> Add AMMClawback Transaction (XLS-0073d) (#5142)
> ```
> 
> ?
> 
> If not, please suggest the details you'd like to include.

I think it's fine that we keep the current message
